### PR TITLE
(! URGENT !) Refactoring: Implemented Factory Method for TLClassifier

### PR DESCRIPTION
**Pull request by ser94mor**
Please, follow the provided structure when creating your classifiers. 
@Elgeweily, please, adopt your PR #8 for this structure (once this PR is merged to _master_)

I have spent a tremendous amount of time trying to adopt the unit test for the _rostest_ or _rosunit_ to be able to execute unit tests with `catkin_make run_tests` but failed to do this. You can run the provided test with `pytest`. It will work, but it does not comply with ROS standards. I would be glad if you suggest me how to integrate this unit test into _catkin_.

### CODE REVIEW

File: ros/src/tl_detector/light_classification/tl_classifier.py [hunk 1]
```diff
@@ -1,78 +1,160 @@
...
+    def __init__(self, cls_name):
+        """
+        Constructor is marked as @abstractmethod to force implemnting the __init__ method in subclasses.
+        Subclasses must invoke their parent constructors.
+        :param cls_name: string identifier of the subclass.
+        """
+        rospy.loginfo("instantiating %s (available classifiers: %s)",
+                      cls_name, str(self.KNOWN_TRAFFIC_LIGHT_CLASSIFIERS.keys()))
+
+        self._lock = threading.Lock()  # lock to be used in TLClassifier.classify to increment invocation counter
```
Comment (thejerrytan) [inline ros/src/tl_detector/light_classification/tl_classifier.py:120]: why do u need locks and threads here?

File: ros/src/tl_detector/light_classification/tl_classifier.py [hunk 2]
```diff
@@ -1,78 +1,160 @@
...
+        Prints FPS statistic approximately each second.
+
+        :param image: image containing the traffic light
+        :type image: np.ndarray
+        :returns: ID of traffic light color (specified in styx_msgs/TrafficLight)
+        :rtype: int
+        """
+        # calculate FPS based on the number of images processed per second;
+        # ensure that self._counter value does not go over integer limits
+        with self._lock:
```
Comment (thejerrytan) [inline ros/src/tl_detector/light_classification/tl_classifier.py:89]: why do u need to acquire a lock before you can run the code?  Is this blocking wait?

**Review by thejerrytan — CHANGES_REQUESTED** [review]

**Review by pkorivi — APPROVED** [review]
It looks good for me, I am not very good at using OOPs concepts in Python, so it's a bit less experienced look.

File: ros/src/tl_detector/light_classification/tl_classifier.py (hunk 2 above)
Comment (thejerrytan) [reply ros/src/tl_detector/light_classification/tl_classifier.py:89]: usually you will need locks if there are multiple threads trying to access a shared resource, i haven't  used threads in python for a long time, so ...

**Comment by thejerrytan** [comment]
Some of the code here is beyond my understanding of python at the moment. Like I am not sure how to decorators work. The other stuff like cls methods and inheritance seems correct.

But so far LGTM.

**Comment by thejerrytan** [comment]
i tried running the code without acquiring locks, surprisingly, the FPS dropped from 10 to 7.

File: ros/src/tl_detector/light_classification/tl_classifier.py (hunk 2 above)
Comment (ser94mor) [reply ros/src/tl_detector/light_classification/tl_classifier.py:89]: TLClassifier.classify is called from the callback of some Subscriber in TLDetector. Callback may be called in parallel by ROS subscribers. `+=` is not an atomic operation in Python, thus without lock self._counter may miss some of the increments. Yes, it is a blocking wait. Critical section has 1 condition check, one increment and 2 assignments, so, threads do not spend much time in the critical section.

File: ros/src/tl_detector/light_classification/tl_classifier.py (hunk 1 above)
Comment (ser94mor) [reply ros/src/tl_detector/light_classification/tl_classifier.py:120]: Yes, see my another comment for explanation.

**Review by thejerrytan — APPROVED** [review]

**Comment by ser94mor** [comment]
Please, comment more if you have questions. I will address your comments in the next PR.

**Pull request merged**